patched Makefile to use prepackaged Google binary of go

crankycoder opened the pull request:
Uses go 1.1.1 binary release instead of checking out and building our own binary.

This fixes https://github.com/mozilla-services/heka/issues/180

crankycoder committed:
patched Makefile to use prepackaged Google binary of go

crankycoder committed:
patched Makefile to continue using build/go if it exists, otherwise we
try to use a release binary

crankycoder committed:
added a check_goroot python script to look for go in some common directories
changed the GOBIN target to only attempt building the hg checkout of go if heka-build/build/go exists, otherwise, try using the google prepackaged go binary

crankycoder commented on Makefile:
```diff
@@ -54,18 +57,31 @@ docs: $(HERE)/heka-docs $(HERE)/pythonVE/bin/sphinx-build bin/hekad
 
 build/go:
-	if [ ! -f $(HGBIN) ]; \
+	if [ ! -f $(GOROOT)/bin/go ]; \
```
I'm only conditionally checkout out the mercurial version of go if the GOROOT is defined, but no GOROOT/bin/go exists.

rafrombrc replied:
So what happens if GOROOT is defined as something outside of heka-build (e.g. GOROOT=/usr/lib/go), but there is no $GOROOT/bin/go file. We install go ourselves and ignore the specified GOROOT?

crankycoder commented on Makefile:
```diff
@@ -83,7 +99,7 @@ heka-source: src/github.com/mozilla-services/heka/README.md
 	GOPATH=$GOPATH PATH="$(HERE)/pythonVE/bin:$(PATH)" python scripts/update_deps.py package_deps.txt
 	@cd src && \
-		$(GOCMD) install github.com/mozilla-services/heka/cmd/hekad
+		$(GOCMD) install -ldflags="-linkmode=external"  github.com/mozilla-services/heka/cmd/hekad
```
linkmode=external works around the bug in https://code.google.com/p/go/issues/detail?id=5726

crankycoder committed:
updated check_goroot to make stricter checks for a GOROOT

crankycoder committed:
got rid of trying to build go from source and always copy over the
latest go binary

crankycoder commented:
@rafrombrc i've gotten rid of the attempts to try building go and added version checks to make sure we've got at least go 1.1

rafrombrc commented on Makefile:
```diff
@@ -53,19 +56,8 @@ docs: $(HERE)/heka-docs $(HERE)/pythonVE/bin/sphinx-build bin/hekad
-	PATH="$(BIN):$(HERE)/pythonVE/bin:$(PATH)" ./all.bash
-	cp build/go/bin/go $(HERE)/bin/go
+$(GOBIN): FORCE check_goroot
+	cp $(GOROOT)/bin/go $(HERE)/bin/go
```
Hrm. I think it's probably better for us to not copy the go binary but to just start running it directly from where it lives. Otherwise if someone updates their system go installation but doesn't run make they'll still be running the older copy of go. How much work would it be to adjust the rest of the build to use go wherever we find it instead of copying it into here?

rafrombrc commented on scripts/check_goroot.py:
```diff
@@ -0,0 +1,86 @@
+
+VALID_GOROOTS = ['/usr/lib/go',
+                 '/usr/local/go',
+                 ]
```
We still want to have '$HERE/build/go' as a valid GOROOT so folks who have a go that an earlier version of heka-build created can still use it.

rafrombrc committed:
Don't barf on non-digits in version string (e.g. "1.1rc3"), and use python's version
compare utilities.

rafrombrc committed:
Check heka-build directory for go installation.

rafrombrc committed:
`check_goroot.py` now either prints a >=1.1 GOROOT path, if it finds one, or nothing,
if not. `Makefile` now sets GOROOT and GOBIN values accordingly, errors out if this
can't be done.

rafrombrc commented:
Okay, tweaked it so it will set the GOROOT if it finds a good one, or raise an error if it can't. Existing checkouts w/ a locally built go installation should continue to work.
